Fix x86 InlinedCallFrame popping during EH

janvorli opened the pull request:
There is a bug in InlinedCallFrame popping on x86 during EH with R2R
compiled code. The code is popping this frame if it is at the top of
the explicit frames stack before executing catch handler. But it
is missing a check to see if the frame was actually in the unwound part
of the stack. So in some corner cases, it can end up popping a frame
that belongs to an active pinvoke somewhere up the call chain.

The fix is to do the popping only if the InlinedCallFrame is located at
an address smaller than the resume ESP.

Close #2240

janvorli committed:
Fix x86 InlinedCallFrame popping during EH

There is a bug in InlinedCallFrame popping on x86 during EH with R2R
compiled code. The code is popping this frame if it is at the top of
the explicit frames stack before executing catch handler. But it
is missing a check to see if the frame was actually in the unwound part
of the stack. So in some corner cases, it can end up popping a frame
that belongs to an active pinvoke somewhere up the call chain.

The fix is to do the popping only if the InlinedCallFrame is located at
an address smaller than the resume ESP.

janvorli commented:
Fixes a problem blocking PR #1929. I've ran corefx tests on Windows x86 with this fix applied on top of the changes from that PR and they have passed.

cc: @clamp03.

jkotas commented on src/coreclr/src/vm/i386/excepx86.cpp:
```diff
@@ -1868,8 +1868,10 @@ NOINLINE LPVOID COMPlusEndCatchWorker(Thread * pThread)
+    if ((pThread->m_pFrame < esp) && InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
     {
+        // The current explicit frame is below the resume esp and it is an inlined call frame with an active call.
+        //
```
Nit: You can fix typos in the comment below while you are on it:

unlined 
pinovoke

janvorli committed:
Fix typos in a comment close to my change

jkotas commented on src/coreclr/src/vm/i386/excepx86.cpp:
```diff
@@ -1868,8 +1868,10 @@ NOINLINE LPVOID COMPlusEndCatchWorker(Thread * pThread)
     pThread->SyncManagedExceptionState(fIsDebuggerHelperThread);
 
-    if (InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
+    if ((pThread->m_pFrame < esp) && InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
```
How does this work?

I believe that we need to check whether the frame is within the method that we are resuming into, ie we need to check whether `pFrame < callerESP`.

janvorli replied:
Hmm, you are right that this would not work for the case when the location we resume to is in the same function as the pinvoke frame. It works when it is anywhere up the call chain.

jkotas replied:
I would not expect the current version of the change to fix the problem.

Anyway, this is not a new problem. #1929 was just unlucky to hit it. I have opened #2240 to track it and added workaround to #1929 to make forward progress.

janvorli committed:
Use correct ESP to compare to the location of the InlinedCallFrame

The correct ESP to use for detecting the InlinedCallFrame affected by
the current exception handling is the the ESP of the caller of the
method that contains the exception handler.

I have moved the check to a different location where the CrawlFrame with
the context of the exception handler is still available and we can
unwind it to get the caller's SP.

jkotas commented:
Could you please also delete the workaround for this bug from AssemblyNative::InternalLoad ?

janvorli committed:
Remove workaround for the original issue

janvorli commented:
@jkotas I've removed the workaround for the issue, can you please take a look again?

jkotas reviewed (approved):
Thanks!
